cql3/expr/expression.cc

```diff
@@ -2349,5 +2349,67 @@ const column_value& get_the_only_column(const expression& e) {
 
     return *result;
 }
+
+
+bool schema_pos_column_definition_comparator::operator()(const column_definition *def1, const column_definition *def2) const {
+    auto column_pos = [](const column_definition* cdef) -> uint32_t {
+        if (cdef->is_primary_key()) {
+            return cdef->id;
+        }
+
+        return 123456789;
+    };
+
+    uint32_t pos1 = column_pos(def1);
+    uint32_t pos2 = column_pos(def2);
+    if (pos1 != pos2) {
+        return pos1 < pos2;
+    }
+    // FIXME: shouldn't we use regular column name comparator here? Origin does not...
+    return less_unsigned(def1->name(), def2->name());
+}
+
+std::vector<const column_definition*> get_sorted_column_defs(const expression& e) {
+    std::set<const column_definition*, schema_pos_column_definition_comparator> cols;
+
+    find_in_expression<column_value>(e,
+        [&](const column_value& cval) -> bool {
+            cols.insert(cval.col);
```

Comment:
What if it's already present?

Reply:
It's a set, so it will deduplicate automatically.